diff --git a/Backends/System/Android/Java-Sources/tech/kinc/KincActivity.java b/Backends/System/Android/Java-Sources/tech/kinc/KincActivity.java
index 6678b2e7..819c6455 100644
--- a/Backends/System/Android/Java-Sources/tech/kinc/KincActivity.java
+++ b/Backends/System/Android/Java-Sources/tech/kinc/KincActivity.java
@@ -13,11 +13,16 @@ import android.os.Message;
 import android.os.Vibrator;
 import android.os.VibrationEffect;
 import android.os.Build;
+import android.view.DragAndDropPermissions;
+import android.view.DragEvent;
 import android.view.KeyEvent;
 import android.view.View;
 import android.view.WindowManager;
 import android.view.inputmethod.InputMethodManager;
 
+import static android.view.DragEvent.ACTION_DRAG_STARTED;
+import static android.view.DragEvent.ACTION_DROP;
+
 public class KincActivity extends NativeActivity implements KeyEvent.Callback {
 	private static KincActivity instance;
 	private InputMethodManager inputManager;
@@ -46,6 +51,24 @@ public class KincActivity extends NativeActivity implements KeyEvent.Callback {
         } catch (PackageManager.NameNotFoundException|NullPointerException e) {
             isDisabledStickyImmersiveMode = false;
         }
+
+		getWindow().getDecorView().setOnDragListener(
+			new View.OnDragListener() {
+				@Override
+				public boolean onDrag(View view, DragEvent dragEvent) {
+					switch(dragEvent.getAction()) {
+						case ACTION_DRAG_STARTED:
+							return true;
+						case ACTION_DROP:
+							DragAndDropPermissions dropPermissions = requestDragAndDropPermissions(dragEvent);
+							importFile(dragEvent.getClipData().getItemAt(0).getUri());
+							dropPermissions.release();
+							return true;
+					}
+					return false;
+				}
+			}
+		);
 	}
 
     private void hideSystemUI() {
@@ -157,4 +180,29 @@ public class KincActivity extends NativeActivity implements KeyEvent.Callback {
 	}
 
 	private native void nativeKincKeyPress(String chars);
+
+	private native void onAndroidFilePicked(String pickedPath);
+
+	@Override
+	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+		super.onActivityResult(requestCode, resultCode, data);
+		if (requestCode == 1 && resultCode == RESULT_OK && data != null) {
+			importFile(data.getData());
+		}
+	}
+
+	public static void pickFile() {
+		Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
+		intent.setType("*/*");
+		getInstance().startActivityForResult(Intent.createChooser(intent, "Select File"), 1);
+	}
+
+	private void importFile(Uri pickedFile) {
+		String[] filePath = { android.provider.MediaStore.Images.Media.DATA };
+		android.database.Cursor cursor = getContentResolver().query(pickedFile, filePath, null, null, null);
+		cursor.moveToFirst();
+		String pickedPath = cursor.getString(cursor.getColumnIndex(filePath[0]));
+		cursor.close();
+		onAndroidFilePicked(pickedPath);
+	}
 }
