diff --git a/Sources/kinc/image.c b/Sources/kinc/image.c
index e5a8bb28..ab7a9726 100644
--- a/Sources/kinc/image.c
+++ b/Sources/kinc/image.c
@@ -63,9 +63,11 @@ static void *buffer_realloc(void *p, size_t size) {
 
 static void buffer_free(void *p) {}
 
+#ifndef KINC_IMAGE_STANDARD_MALLOC
 #define STBI_MALLOC(sz) buffer_malloc(sz)
 #define STBI_REALLOC(p, newsz) buffer_realloc(p, newsz)
 #define STBI_FREE(p) buffer_free(p)
+#endif
 
 #define STB_IMAGE_IMPLEMENTATION
 #include <kinc/libs/stb_image.h>
