diff --git a/Sources/kinc/graphics5/raytrace.h b/Sources/kinc/graphics5/raytrace.h
index fdba1c8b..bf116676 100644
--- a/Sources/kinc/graphics5/raytrace.h
+++ b/Sources/kinc/graphics5/raytrace.h
@@ -20,14 +20,14 @@ typedef struct kinc_raytrace_acceleration_structure {
 	kinc_raytrace_acceleration_structure_impl_t impl;
 } kinc_raytrace_acceleration_structure_t;

-void kinc_raytrace_acceleration_structure_init(kinc_raytrace_acceleration_structure_t *accel, kinc_g5_command_list_t *command_list, kinc_g5_vertex_buffer_t *vb, kinc_g5_index_buffer_t *ib);
+void kinc_raytrace_acceleration_structure_init(kinc_raytrace_acceleration_structure_t *accel, kinc_g5_command_list_t *command_list, kinc_g5_vertex_buffer_t *vb, kinc_g5_index_buffer_t *ib, float scale);
 void kinc_raytrace_acceleration_structure_destroy(kinc_raytrace_acceleration_structure_t *accel);

+void kinc_raytrace_set_textures(kinc_g5_render_target_t* texpaint0, kinc_g5_render_target_t *texpaint1, kinc_g5_render_target_t *texpaint2, kinc_g5_texture_t *texenv, kinc_g5_texture_t *texsobol, kinc_g5_texture_t *texscramble, kinc_g5_texture_t *texrank);
 void kinc_raytrace_set_acceleration_structure(kinc_raytrace_acceleration_structure_t *accel);
 void kinc_raytrace_set_pipeline(kinc_raytrace_pipeline_t *pipeline);
-void kinc_raytrace_set_target(kinc_g5_texture_t *output);
+void kinc_raytrace_set_target(kinc_g5_render_target_t *output);
 void kinc_raytrace_dispatch_rays(kinc_g5_command_list_t *command_list);
-void kinc_raytrace_copy(kinc_g5_command_list_t *command_list, kinc_g5_render_target_t *target, kinc_g5_texture_t *source);

 #ifdef __cplusplus
 }
