diff --git a/Backends/System/Windows/Sources/kinc/backend/system.cpp b/Backends/System/Windows/Sources/kinc/backend/system.cpp
index 2089033a..135cd61c 100644
--- a/Backends/System/Windows/Sources/kinc/backend/system.cpp
+++ b/Backends/System/Windows/Sources/kinc/backend/system.cpp
@@ -246,6 +246,8 @@ static bool detectGamepad = true;
 static bool gamepadFound = false;
 static unsigned r = 0;

+void armory_save_and_quit(bool save);
+
 namespace {
 	wchar_t toUnicode(WPARAM wParam, LPARAM lParam) {
 		wchar_t buffer[11];
@@ -296,6 +298,17 @@ extern "C" LRESULT WINAPI KoreWindowsMessageProcedure(HWND hWnd, UINT msg, WPARA
 		}
 		break;
 	}
+	case WM_CLOSE: {
+		int res = IDNO;
+		wchar_t title[1024];
+		GetWindowTextW(hWnd, title, sizeof(title));
+		bool dirty = wcsstr(title, L"* - ArmorPaint") != NULL;
+		if (dirty) {
+			res = MessageBox(hWnd, L"Project has been modified, save changes?", L"Save Changes?", MB_YESNOCANCEL | MB_ICONEXCLAMATION );
+		}
+		if (res != IDCANCEL) armory_save_and_quit(res == IDYES);
+		return 0;
+	}
 	case WM_DESTROY:
 		kinc_stop();
 		return 0;
diff --git a/Backends/System/Windows/Sources/kinc/backend/window.c b/Backends/System/Windows/Sources/kinc/backend/window.c
index a107a3b3..98cc764f 100644
--- a/Backends/System/Windows/Sources/kinc/backend/window.c
+++ b/Backends/System/Windows/Sources/kinc/backend/window.c
@@ -205,8 +205,10 @@ static int createWindow(const wchar_t *title, int x, int y, int width, int heigh

 	switch (windowMode) {
 	case KINC_WINDOW_MODE_WINDOW:
-		dstx += x < 0 ? (display_mode.width - width) / 2 : x;
-		dsty += y < 0 ? (display_mode.height - height) / 2 : y;
+		// Maximized window
+		// Do not center if window width (including borders) exceeds display width
+		dstx += (x < 0 && width < display_mode.width) ? (display_mode.width - width) / 2 : x;
+		dsty += (y < 0 && width < display_mode.width) ? (display_mode.height - height) / 2 : y;
 		dstw = WindowRect.right - WindowRect.left;
 		dsth = WindowRect.bottom - WindowRect.top;
 		break;
