diff --git a/Backends/System/Windows/Sources/kinc/backend/window.c.h b/Backends/System/Windows/Sources/kinc/backend/window.c.h
index a1393a88..cac7cef7 100644
--- a/Backends/System/Windows/Sources/kinc/backend/window.c.h
+++ b/Backends/System/Windows/Sources/kinc/backend/window.c.h
@@ -225,8 +225,10 @@ static int createWindow(const wchar_t *title, int x, int y, int width, int heigh
 
 	switch (windowMode) {
 	case KINC_WINDOW_MODE_WINDOW:
-		dstx += x < 0 ? (display_mode.width - width) / 2 : x;
-		dsty += y < 0 ? (display_mode.height - height) / 2 : y;
+		// Maximized window
+		// Do not center if window width (including borders) exceeds display width
+		dstx += (x < 0 && width < display_mode.width) ? (display_mode.width - width) / 2 : x;
+		dsty += (y < 0 && width < display_mode.width) ? (display_mode.height - height) / 2 : y;
 		dstw = WindowRect.right - WindowRect.left;
 		dsth = WindowRect.bottom - WindowRect.top;
 		break;
