diff --git a/Backends/System/iOS/Sources/kinc/backend/GLView.m.h b/Backends/System/iOS/Sources/kinc/backend/GLView.m.h
index bdf70950..b3898fe1 100644
--- a/Backends/System/iOS/Sources/kinc/backend/GLView.m.h
+++ b/Backends/System/iOS/Sources/kinc/backend/GLView.m.h
@@ -80,6 +80,14 @@ extern int kinc_ios_gl_framebuffer;
 #ifdef KORE_METAL
 void initMetalCompute(id<MTLDevice> device, id<MTLCommandQueue> commandQueue);
 
+- (void)hoverGesture:(UIHoverGestureRecognizer *)recognizer {
+	CGPoint point = [recognizer locationInView:self];
+	float x = point.x * self.contentScaleFactor;
+	float y = point.y * self.contentScaleFactor;
+	// Pencil hover
+	kinc_internal_pen_trigger_move(0, x, y, 0.0);
+}
+
 - (id)initWithFrame:(CGRect)frame {
 	self = [super initWithFrame:(CGRect)frame];
 	self.contentScaleFactor = [UIScreen mainScreen].scale;
@@ -104,6 +112,7 @@ void initMetalCompute(id<MTLDevice> device, id<MTLCommandQueue> commandQueue);
 	metalLayer.opaque = YES;
 	metalLayer.backgroundColor = nil;
 
+	[self addGestureRecognizer:[[UIHoverGestureRecognizer alloc] initWithTarget:self action:@selector(hoverGesture:)]];
 	return self;
 }
 #else
